File: modules/cubex-i18n/src/main/java/org/cubexmc/i18n/I18nOptions.java
package org.cubexmc.i18n;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.EnumSet;
import java.util.List;
import java.util.Set;
import java.util.function.Supplier;

public final class I18nOptions {

    private String languageDirectory = "lang";
    private String currentLocale = "zh_CN";
    private Supplier<String> currentLocaleSupplier;
    private String defaultLocale = "zh_CN";
    private final List<String> fallbackLocales = new ArrayList<>();
    private final List<String> bundledLocales = new ArrayList<>();
    private String prefixKey = "prefix";
    private String prefixToken = "%prefix%";
    private String keyPrefix = "";
    private MissingKeyMode missingKeyMode = MissingKeyMode.RETURN_KEY;
    private boolean warnOnMissingKey;
    private Set<PlaceholderStyle> placeholderStyles = EnumSet.of(PlaceholderStyle.PERCENT_NAME);
    private ColorMode colorMode = ColorMode.LEGACY_AND_HEX;

    private I18nOptions() {
    }

    public static I18nOptions create() {
        return new I18nOptions();
    }

    public I18nOptions languageDirectory(String directory) {
        this.languageDirectory = directory == null || directory.isBlank() ? "lang" : directory;
        return this;
    }

    public I18nOptions currentLocale(String locale) {
        this.currentLocale = locale;
        this.currentLocaleSupplier = null;
        return this;
    }

    public I18nOptions currentLocale(Supplier<String> localeSupplier) {
        this.currentLocaleSupplier = localeSupplier;
        return this;
    }

    public I18nOptions defaultLocale(String locale) {
        this.defaultLocale = locale;
        return this;
    }

    public I18nOptions fallbackLocales(List<String> locales) {
        this.fallbackLocales.clear();
        if (locales != null) {
            this.fallbackLocales.addAll(locales);
        }
        return this;
    }

    public I18nOptions bundledLocales(Collection<String> locales) {
        this.bundledLocales.clear();
        if (locales != null) {
            this.bundledLocales.addAll(locales);
        }
        return this;
    }

    public I18nOptions prefixKey(String key) {
        this.prefixKey = key == null ? "" : key;
        return this;
    }

    public I18nOptions prefixToken(String token) {
        this.prefixToken = token == null ? "" : token;
        return this;
    }

    public I18nOptions keyPrefix(String keyPrefix) {
        this.keyPrefix = keyPrefix == null ? "" : keyPrefix;
        return this;
    }

    public I18nOptions missingKeyMode(MissingKeyMode mode) {
        this.missingKeyMode = mode == null ? MissingKeyMode.RETURN_KEY : mode;
        return this;
    }

    public I18nOptions warnOnMissingKey(boolean enabled) {
        this.warnOnMissingKey = enabled;
        return this;
    }

    public I18nOptions placeholderStyles(Collection<PlaceholderStyle> styles) {
        if (styles == null || styles.isEmpty()) {
            this.placeholderStyles = EnumSet.noneOf(PlaceholderStyle.class);
        } else {
            this.placeholderStyles = EnumSet.copyOf(styles);
        }
        return this;
    }

    public I18nOptions colorize(boolean enabled) {
        this.colorMode = enabled ? ColorMode.LEGACY_AND_HEX : null;
        return this;
    }

    public I18nOptions colorMode(ColorMode mode) {
        this.colorMode = mode;
        return this;
    }

    String languageDirectory() {
        return languageDirectory;
    }

    String currentLocale() {
        String supplied = currentLocaleSupplier == null ? null : currentLocaleSupplier.get();
        return supplied == null || supplied.isBlank() ? currentLocale : supplied;
    }

    boolean hasCurrentLocaleSupplier() {
        return currentLocaleSupplier != null;
    }

    String defaultLocale() {
        return defaultLocale;
    }

    List<String> fallbackLocales() {
        return Collections.unmodifiableList(fallbackLocales);
    }

    List<String> bundledLocales() {
        return Collections.unmodifiableList(bundledLocales);
    }

    String prefixKey() {
        return prefixKey;
    }

    String prefixToken() {
        return prefixToken;
    }

    String keyPrefix() {
        return keyPrefix;
    }

    MissingKeyMode missingKeyMode() {
        return missingKeyMode;
    }

    boolean warnOnMissingKey() {
        return warnOnMissingKey;
    }

    Set<PlaceholderStyle> placeholderStyles() {
        return Collections.unmodifiableSet(placeholderStyles);
    }

    ColorMode colorMode() {
        return colorMode;
    }
}

File: modules/cubex-i18n/src/main/kotlin/org/cubexmc/i18n/I18nOptions.kt
package org.cubexmc.i18n

import java.util.EnumSet
import java.util.function.Supplier

class I18nOptions private constructor() {
    private var languageDirectoryValue = "lang"
    private var currentLocaleValue: String? = "zh_CN"
    private var currentLocaleSupplier: Supplier<String>? = null
    private var defaultLocaleValue: String? = "zh_CN"
    private val fallbackLocaleValues = ArrayList<String>()
    private val bundledLocaleValues = ArrayList<String>()
    private var prefixKeyValue = "prefix"
    private var prefixTokenValue = "%prefix%"
    private var keyPrefixValue = ""
    private var missingKeyModeValue = MissingKeyMode.RETURN_KEY
    private var warnOnMissingKeyValue = false
    private var placeholderStyleValues: Set<PlaceholderStyle> = EnumSet.of(PlaceholderStyle.PERCENT_NAME)
    private var colorModeValue: ColorMode? = ColorMode.LEGACY_AND_HEX

    fun languageDirectory(directory: String?): I18nOptions = apply {
        languageDirectoryValue = directory?.takeUnless { it.isBlank() } ?: "lang"
    }

    fun currentLocale(locale: String?): I18nOptions = apply {
        currentLocaleValue = locale
        currentLocaleSupplier = null
    }

    fun currentLocale(localeSupplier: Supplier<String>?): I18nOptions = apply {
        currentLocaleSupplier = localeSupplier
    }

    fun defaultLocale(locale: String?): I18nOptions = apply { defaultLocaleValue = locale }

    fun fallbackLocales(locales: List<String>?): I18nOptions = apply {
        fallbackLocaleValues.clear()
        if (locales != null) fallbackLocaleValues.addAll(locales)
    }

    fun bundledLocales(locales: Collection<String>?): I18nOptions = apply {
        bundledLocaleValues.clear()
        if (locales != null) bundledLocaleValues.addAll(locales)
    }

    fun prefixKey(key: String?): I18nOptions = apply { prefixKeyValue = key ?: "" }
    fun prefixToken(token: String?): I18nOptions = apply { prefixTokenValue = token ?: "" }
    fun keyPrefix(keyPrefix: String?): I18nOptions = apply { keyPrefixValue = keyPrefix ?: "" }
    fun missingKeyMode(mode: MissingKeyMode?): I18nOptions =
        apply { missingKeyModeValue = mode ?: MissingKeyMode.RETURN_KEY }
    fun warnOnMissingKey(enabled: Boolean): I18nOptions = apply { warnOnMissingKeyValue = enabled }

    fun placeholderStyles(styles: Collection<PlaceholderStyle>?): I18nOptions = apply {
        placeholderStyleValues =
            if (styles.isNullOrEmpty()) EnumSet.noneOf(PlaceholderStyle::class.java) else EnumSet.copyOf(styles)
    }

    fun colorize(enabled: Boolean): I18nOptions = apply {
        colorModeValue = if (enabled) ColorMode.LEGACY_AND_HEX else null
    }

    fun colorMode(mode: ColorMode?): I18nOptions = apply { colorModeValue = mode }

    internal fun languageDirectory(): String = languageDirectoryValue
    internal fun currentLocale(): String? =
        currentLocaleSupplier?.get()?.takeUnless { it.isBlank() } ?: currentLocaleValue
    internal fun hasCurrentLocaleSupplier(): Boolean = currentLocaleSupplier != null
    internal fun defaultLocale(): String? = defaultLocaleValue
    internal fun fallbackLocales(): List<String> = fallbackLocaleValues.toList()
    internal fun bundledLocales(): List<String> = bundledLocaleValues.toList()
    internal fun prefixKey(): String = prefixKeyValue
    internal fun prefixToken(): String = prefixTokenValue
    internal fun keyPrefix(): String = keyPrefixValue
    internal fun missingKeyMode(): MissingKeyMode = missingKeyModeValue
    internal fun warnOnMissingKey(): Boolean = warnOnMissingKeyValue
    internal fun placeholderStyles(): Set<PlaceholderStyle> = placeholderStyleValues.toSet()
    internal fun colorMode(): ColorMode? = colorModeValue

    companion object {
        @JvmStatic
        fun create(): I18nOptions = I18nOptions()
    }
}
